### PR TITLE
add Disjoint Groups RFC

### PR DESCRIPTION
**Pull request by afck**
[Rendered](https://github.com/afck/rfcs/blob/master/text/0037-disjoint-groups/0037-disjoint-groups.md)

This change is [Reviewable](https://reviewable.io/reviews/maidsafe/rfcs/160)

### CODE REVIEW

File: text/0000-disjoint-groups/0000-disjoint-groups.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+### Data chains
+
+The current group definition would considerably complicate [data chains](https://github.com/dirvine/data_chain): With disjoint groups, all members of a group can sign the same link block - consising of all the group members -, and are responsible for exactly the same set of data chunks, so they can sign the same data blocks.
+
+
+## Detailed design
+
+The definition of a close group is modified so that at each point in time, the name space is partitioned into disjoint groups. Each group is defined by a _name prefix_, i.e. a sequence of between 0 and 256 bits, similar to how IP subnets correspond to IP address prefixes. The group `(p)` consists of all nodes whose name begins with the prefix `p`, and is responsible for all data items whose name begins with `p`. The group thus manages the part of the name space given by the interval `[p00...00, p11...11]`.
```
Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: Should this read "between 0 and n-1 bits" perhaps.

File: text/0000-disjoint-groups/0000-disjoint-groups.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+* Whenever `(p0)` and `(p1)` satisfy certain group requirements (see below), a group `(p)` splits into `(p0)` and `(p1)`.
+* Whenever a group `(p0)` or `(p1)` ceases to satisfy the group requirements, it merges with all its sister groups into the group `(p)` again: All groups that would be a subset of `(p)` are merged back into `(p)`.
+
+These are the only two operations allowed on groups, so it is guaranteed that every address in the network belongs to exactly one group, and that group satisfies the requirements.
+
+The invariant that needs to be satisfied by the routing table is modified accordingly:
+
+1. A node must have its complete group `(p)` in its routing table.
+2. It must be connected to every group `(q)`, for which `p` and `q` differ in exactly one bit.
```
Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: Should this be "fully connected" or connected to each node, or are we considering and advanced mechanism perhaps to allow proxy connections or similar?

File: text/0000-disjoint-groups/0000-disjoint-groups.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+If `p` and `q` are not of equal length, that does not count as a difference: A differing bit is one that is defined in _both_ prefixes, but is 1 in one of them and 0 in the other. So e. g. `0000` and `001000` differ in exactly one bit.
+
+The groups that differ in the `i`-th bit are the "`i`-th bucket" of `(p)`.
+
+In a balanced network, point 2 just means that the group `(q)` _is_ the `i`-th bucket. But if it is not perfectly balanced, it might be the two groups `(q0)` and `(q1)` or even more groups with longer prefixes, or some prefix of `q` itself.
+
+The requirement is symmetric: I need you in my routing table if and only if you need me in yours!
```
Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: This is a much stronger mechanism as well I feel and perhaps is another advantage over current?

File: text/0000-disjoint-groups/0000-disjoint-groups.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+The quorum cannot be a constant anymore, due to varying group sizes. It needs to be a percentage strictly greater than 50% instead, and in a group of size `n`, a number `x` of nodes will constitute a quorum if `x / n >= QUORUM`.
+
+### Joining nodes
+
+A new node needs to first connect to everyone in its target group. After that, it can drop its proxy node. The other group members can provide the node with their routing tables, which should all be identical. It can then make connections to these other groups, too.
+
+### Group split
+
+If a group `(p)` splits, all its nodes stay connected: The new groups are now each other's `i`-th bucket, where `i` is the number of bits in `p`.
```
Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: Does it also disconnect from the previous connected group that agreed on the current leading bits - 1 ?

File: text/0000-disjoint-groups/0000-disjoint-groups.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+The group sends a direct message to all its contacts to inform them about the split. These need to note that the group has split, even if they stay connected to all its members.
+
+The node can then disconnect immediately (no more `ConnectionUnneeded` required) from every group `(q)` for which `q` now differs in more than one bit from its new group prefix.
```
Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: Ah answered above question, sry for noise.

Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: I think this then needs a Splitting message to vaults, same for converge perhaps. If vaults do have DataChains then aany data on converge again needs to republish (i.e. send all datablocks from the previous group which was CLB (Common Leading Bits) + 1 == our address).

File: text/0000-disjoint-groups/0000-disjoint-groups.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+### Leaving nodes
+
+A leaving node will cause `LostPeer` events in all its connected peers. No further action is necessary, except if it causes a group merge.
+
+### Group merge
+
+When a new group `(p)` is created by merging, all its members are already connected. They need to exchange their routing tables, merge them and establish the missing connections. Finally, they need to notify all of their peers about the group merge. Each node's new routing table will be the union of all the group members' previous tables.
```
Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: Yes so here this message may need to propagate to vault.  This is an extension of previous comment.

File: text/0000-disjoint-groups/0000-disjoint-groups.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+
+## Drawbacks
+
+The main drawback is that, compared to the `GROUP_SIZE` constant, groups and routing tables will become larger. Since `GROUP_SIZE` will turn into a minimum instead of an exact size requirement, and groups will usually (if they are balanced) split when they have about `2 * GROUP_SIZE` members, the average group will probably have at least 50% more than `GROUP_SIZE` members. This increases the entries in the routing table and the number of network connections that are simultaneously open. It also increases traffic whenever _each member_ of a group has to take action, e.g. for group messages or data chain signatures.
```
Comment (dirvine) [inline text/0000-disjoint-groups/0000-disjoint-groups.md]: Here the groups will be come larger proportional to the unbalance of a binary tree perhaps, so will appear worse in smaller networks where the imbalance is more pronounced.

**Comment by afck** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 45 \[r1\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNCoKLGctq2D2BR0h4c-r1-45:-KND0dcWV4B98MmvqyNW:-78195738) ([raw file](https://github.com/maidsafe/rfcs/blob/552c97beb78992e51c4795491f4c80cc21d8ef3e/text/0000-disjoint-groups/0000-disjoint-groups.md#L45)):*
<details><summary><i>Previously, dirvine (David Irvine) wrote…</i></summary>
> Should this read "between 0 and n-1 bits" perhaps.
</details>
Done.
(I used `XOR_NAME_BITS`, as I use `n` elsewhere. Strictly speaking, they are between 0 and `XOR_NAME_BITS - 3` bits, but that doesn't really matter, as they will usually be _much_ shorter anyway.)

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 57 \[r1\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNCoKLGctq2D2BR0h4c-r1-57:-KNCywkx5VkFvBHPzHkJ:-796129014) ([raw file](https://github.com/maidsafe/rfcs/blob/552c97beb78992e51c4795491f4c80cc21d8ef3e/text/0000-disjoint-groups/0000-disjoint-groups.md#L57)):*
<details><summary><i>Previously, dirvine (David Irvine) wrote…</i></summary>
> Should this be "fully connected" or connected to each node, or are we considering and advanced mechanism perhaps to allow proxy connections or similar? 
</details>
It means to every node of the group, yes.
I also changed the formulation to "have them in their routing table" instead of "connected": Tunnel connections are treated the same way as direct connections in this PR.

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 65 \[r1\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNCoKLGctq2D2BR0h4c-r1-65:-KND-5OrRgEf-fUtJln7:-1399286082) ([raw file](https://github.com/maidsafe/rfcs/blob/552c97beb78992e51c4795491f4c80cc21d8ef3e/text/0000-disjoint-groups/0000-disjoint-groups.md#L65)):*
<details><summary><i>Previously, dirvine (David Irvine) wrote…</i></summary>
> This is a much stronger mechanism as well I feel and perhaps is another advantage over current? 
</details>
I agree. Just didn't want to make the "Motivation" section even more verbose. :-)

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 120 \[r1\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNCoKLGctq2D2BR0h4c-r1-120:-KND-_XP79TSEd1Ycs34:-499146009) ([raw file](https://github.com/maidsafe/rfcs/blob/552c97beb78992e51c4795491f4c80cc21d8ef3e/text/0000-disjoint-groups/0000-disjoint-groups.md#L120)):*
<details><summary><i>Previously, dirvine (David Irvine) wrote…</i></summary>
> I think this then needs a Splitting message to vaults, same for converge perhaps. If vaults do have DataChains then aany data on converge again needs to republish (i.e. send all datablocks from the previous group which was CLB (Common Leading Bits) + 1 == our address). 
</details>
Done.

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 128 \[r1\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNCoKLGctq2D2BR0h4c-r1-128:-KND04bxDaBruoKvUbnP:-499146009) ([raw file](https://github.com/maidsafe/rfcs/blob/552c97beb78992e51c4795491f4c80cc21d8ef3e/text/0000-disjoint-groups/0000-disjoint-groups.md#L128)):*
<details><summary><i>Previously, dirvine (David Irvine) wrote…</i></summary>
> Yes so here this message may need to propagate to vault.  This is an extension of previous comment. 
</details>
Done.

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 133 \[r1\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNCoKLGctq2D2BR0h4c-r1-133:-KND0ARjP-9Iz1msx-Dr:858641217) ([raw file](https://github.com/maidsafe/rfcs/blob/552c97beb78992e51c4795491f4c80cc21d8ef3e/text/0000-disjoint-groups/0000-disjoint-groups.md#L133)):*
<details><summary><i>Previously, dirvine (David Irvine) wrote…</i></summary>
> Here the groups will be come larger proportional to the unbalance of a binary tree perhaps, so will appear worse in smaller networks where the imbalance is more pronounced. 
</details>
I don't think the network size even makes much of a difference: In terms of balance and group splits, every new group will behave in exactly the same way as the original network as a whole.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by dirvine** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 153 \[r2\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KND3T0u1E1EG6MU6ZSh:-KND3T0u1E1EG6MU6ZSi:1080783049) ([raw file](https://github.com/maidsafe/rfcs/blob/adec4b235c312520e99f5cbeb1951853d7f1021e/text/0000-disjoint-groups/0000-disjoint-groups.md#L153)):*
> ```Markdown
> ## Unresolved questions
> 
> How will safe_vault handle merging groups and restore data redundancy?
> ```

On receipt of a "Converge" message a vault should publish (send to group) it's Block's that represent data form the part of the chain that just converged (ie current leading bits + 1 (or 0)). So the part of the chain from the current leading bits to the end of it's chain.
This effectively rewrites the chain from the current chekpoint again. Nodes do not need to  delete blocks etc. (as they already have them) but they will add newly seen blocks (form the other prev split group). As the DataChain knows it can accumulate on GROUP_SIZE then this works itself, however both groups can merge / accept the other groups chains and consequently their blocks (as can any network connected node). 
So it would suffice that on coverage each node form each split sends it's chain to the other nodes.  The archive nodes and full nodes will pick up the data from there. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by afck** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[text/0000-disjoint-groups/0000-disjoint-groups.md, line 153 \[r2\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KND3T0u1E1EG6MU6ZSh:-KND9Eb1wBt71BPdoVsE:1501161356) ([raw file](https://github.com/maidsafe/rfcs/blob/adec4b235c312520e99f5cbeb1951853d7f1021e/text/0000-disjoint-groups/0000-disjoint-groups.md#L153)):*
<details><summary><i>Previously, dirvine (David Irvine) wrote…</i></summary>
> On receipt of a "Converge" message a vault should publish (send to group) it's Block's that represent data form the part of the chain that just converged (ie current leading bits + 1 (or 0)). So the part of the chain from the current leading bits to the end of it's chain.
> This effectively rewrites the chain from the current chekpoint again. Nodes do not need to  delete blocks etc. (as they already have them) but they will add newly seen blocks (form the other prev split group). As the DataChain knows it can accumulate on GROUP_SIZE then this works itself, however both groups can merge / accept the other groups chains and consequently their blocks (as can any network connected node). 
> So it would suffice that on coverage each node form each split sends it's chain to the other nodes.  The archive nodes and full nodes will pick up the data from there. 
</details>
I removed the "unsolved question" and added a remark in the last "Drawbacks" paragraph. Ultimately, this group requirement change will probably be part of the later Data Chain introduction (or even a separate issue after they have been introduced), not of the Disjoint Groups.

But we'll still need to be careful when we formulate the exact requirement later: It cannot simply be that three nodes must be "complete", in the sense that they have all the group's data, because a group merge can turn complete nodes into incomplete ones, causing another group merge etc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by dirvine** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-:-KNDBXEkDNZ06nyb6M-M:980077401)*

**Comment by madadam** [comment]
*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 88 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNG_AsiqS5wgTD5Kpq9:-KNG_AsjPAUKCUeNs8cO:223845100) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L88)):*
> ```Markdown
> If the destination is a _group authority_ with address `d`:
> 
> * If `t` extends `p`, handle the message and relay it to everyone else in `(p)`.
> ```

1. What is `t`? Shouldn't it be `d`?
2. What does "`t` extends `p`" mean?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by afck** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 88 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNG_AsiqS5wgTD5Kpq9:-KNGfHpCt0nnrisWgr3W:680522321) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L88)):*
<details><summary><i>Previously, madadam (Adam Cigánek) wrote…</i></summary>
> 1. What is `t`? Shouldn't it be `d`?
> 2. What does "`t` extends `p`" mean?
</details>
Ah, sorry, I had renamed that! I also made the formulation more clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by maqi** [comment]
*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 59 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGhEcATwBmEFrBYu7T:-KNGhEcATwBmEFrBYu7U:-1876497590) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L59)):*
> ```Markdown
> 2. It must have every member of every group `(q)` in its routing table, for which `p` and `q` differ in exactly one bit.
> 
> If `p` and `q` are not of equal length, that does not count as a difference: A differing bit is one that is defined in _both_ prefixes, but is 1 in one of them and 0 in the other. So e. g. `0000` and `001000` differ in exactly one bit.
> ```

so p00 and p01 knows each other, and both of them knows about p1 as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by maqi** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 36 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGfBeVdhKiF0I76SO7:-KNGfBeW-PtbBMuogvKJ:145797188) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L36)):*
> ```Markdown
> ### Weigthed votes and node evaluation
> 
> If group `A` is connected to group `B`, every node in `B` will know every member of `A`, so group `B` can easily reach a consensus regarding the correctness of the individual `A`-members' behaviour. This allows `B` to evaluate the nodes in `A` and give feedback about them - another requirement for banning misbehaving nodes. But it could be taken one step further and used to assign grades, and make a nodes' vote weight for group consensus depend on it: Making voting rights depend on a proof of work is essential for the security of the network.
> ```

will all nodes in group `A` also knows all nodes in group `B`?

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 50 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGfrE1iFVlTrxDi6yN:-KNGfrE1iFVlTrxDi6yO:1751549008) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L50)):*
> ```Markdown
> 
> * Whenever `(p0)` and `(p1)` satisfy certain group requirements (see below), a group `(p)` splits into `(p0)` and `(p1)`.
> * Whenever a group `(p0)` or `(p1)` ceases to satisfy the group requirements, it merges with all its sister groups into the group `(p)` again: All groups that would be a subset of `(p)` are merged back into `(p)`.
> ```

say p has been split into p0 and p1, then p0 got further split into p00 and p01
if p00 doesn't fulfill the requirements anymore, p00 and p01 will merge back to p0, if still not fulfill, then p0 and p1 merge back to p?

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 63 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGkDQyl0SySLUbd1OM:-KNGkDQyl0SySLUbd1ON:-1514159907) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L63)):*
> ```Markdown
> The groups that differ in the `i`-th bit are the "`i`-th bucket" of `(p)`.
> 
> In a balanced network, point 2 just means that the group `(q)` _is_ the `i`-th bucket. But if it is not perfectly balanced, it might be the two groups `(q0)` and `(q1)` or even more groups with longer prefixes, or some prefix of `q` itself.
> ```

which means p1 will always know all sub groups divided from p0 (p00 p010 p0111, p01100 ...) ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by maqi** [comment]
*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 112 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGkYRS9_dYgaYgcQ4y:-KNGkYRTzvyez0r-9Gmn:-332758959) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L112)):*
> ```Markdown
> ### Joining nodes
> 
> A new node needs to first connect to everyone in its target group. After that, it can drop its proxy node. The other group members can provide the node with their routing tables, which should all be identical. It can then make connections to these other groups, too.
> ```

then the info of this new joined node shall be forwarded to all other groups that shall know this group?
and same to the situation of dropping nodes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by afck** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 36 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGfBeVdhKiF0I76SO7:-KNGnGT_qzK9ezhqR5G_:-2098702088) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L36)):*
<details><summary><i>Previously, maqi wrote…</i></summary>
> will all nodes in group `A` also knows all nodes in group `B`?
</details>
Yes: Two groups will either be connected in the sense that each node in each group knows each node in the other group, or disconnected in the sense that none of their nodes have connections to each other.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 50 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGfrE1iFVlTrxDi6yN:-KNGnVCCV9DDCTaAILxS:1850230924) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L50)):*
<details><summary><i>Previously, maqi wrote…</i></summary>
> say p has been split into p0 and p1, then p0 got further split into p00 and p01
> if p00 doesn't fulfill the requirements anymore, p00 and p01 will merge back to p0, if still not fulfill, then p0 and p1 merge back to p?
</details>
Exactly. I added an example.

This is also why we need to be careful if we use more complex group requirements: We need to make sure that after the merge, it is _very likely_ that `p0` will fulfil them again, so that we don't cause a chain reaction that causes the whole network to merge. :-D

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 59 \[r5\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGhEcATwBmEFrBYu7T:-KNGlvZZNFJM8iNIEtdn:-1852533611) ([raw file](https://github.com/maidsafe/rfcs/blob/3772c3e086c676b829cb495de5e13f9ea9f56e66/text/0037-disjoint-groups/0037-disjoint-groups.md#L59)):*
<details><summary><i>Previously, maqi wrote…</i></summary>
> so p00 and p01 knows each other, and both of them knows about p1 as well?
</details>
Yes. Assume e.g. that `p` is `11`. Then those three groups are:
```
1100
1101
111
```
Since any two of them differ in exactly one bit, they need to be connected to each other.

I extended the example in the text a bit.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 63 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGkDQyl0SySLUbd1OM:-KNGlLR1kFAHIpiDRvwa:-1832301044) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L63)):*
<details><summary><i>Previously, maqi wrote…</i></summary>
> which means p1 will always know all sub groups divided from p0 (p00 p010 p0111, p01100 ...) ?
</details>
Yes, that's true, but for `p1` to stay a group and the others to split further and further, the network would need to become _very_ unbalanced.
We're currently assuming that the node names will spread somewhat uniformly across the name space, but if that turns out to be wrong, we could take additional measures to ensure it when assigning node names (or name ranges).

And if `p1` splits into `p10` and `p11`, then `p11` can disconnect from `p00`, and `p10` can disconnect from `p01`.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 112 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGkYRS9_dYgaYgcQ4y:-KNGmNA_rfz7smS-ulOR:2095856010) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L112)):*
<details><summary><i>Previously, maqi wrote…</i></summary>
> then the info of this new joined node shall be forwarded to all other groups that shall know this group?
> and same to the situation of dropping nodes?
</details>
Yes, when we introduce data chains, this will probably mean that for a joining node, its group will forward the link block including the new node, signed by the whole group, to all the connected groups, so that they will accept the connection to the new node.

Not sure about leaving nodes, but you're probably right: Once a node disconnects from me, I probably need to ask its group whether it has actually left the network; otherwise I need to reconnect, or look for a tunnel to it.

I added two paragraphs to the text about this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by Fraser999** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 32 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGjufmpvpflXNTrFHL:-KNGjufnJNTsmAYwq8Fu:-62372861) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L32)):*
> disconnect from that it.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 34 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGiaS0UHD0Y6xJR9Nw:-KNGiaS0UHD0Y6xJR9Nx:-484329001) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L34)):*
> ```Markdown
> The routing tables of all nodes in a group will be identical. This allows the network to ban misbehaving nodes: If a node's group decides to ban the node from the network, it knows exactly whom this node is connected to, and who needs to be notified to make the whole network disconnect from that it.
> 
> ### Weigthed votes and node evaluation
> ```

Tpyo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Comment by afck** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 32 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGjufmpvpflXNTrFHL:-KNGwLG7dpcOjBPOqdya:-499146009) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L32)):*
<details><summary><i>Previously, Fraser999 (Fraser Hutchison) wrote…</i></summary>
> > disconnect from that it.
</details>
Done.

---

*[text/0037-disjoint-groups/0037-disjoint-groups.md, line 34 \[r6\]](https://reviewable.io:443/reviews/maidsafe/rfcs/160#-KNGiaS0UHD0Y6xJR9Nw:-KNGwMDJvRbwxfXnr2Nx:-489920409) ([raw file](https://github.com/maidsafe/rfcs/blob/35594eba2dc78ec5255168660fab9f7d910a744c/text/0037-disjoint-groups/0037-disjoint-groups.md#L34)):*
<details><summary><i>Previously, Fraser999 (Fraser Hutchison) wrote…</i></summary>
> Tpyo.
</details>
Node.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/rfcs/160)*

**Pull request merged**